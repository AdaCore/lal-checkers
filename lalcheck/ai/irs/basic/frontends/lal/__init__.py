"""
Provides a libadalang frontend for the Basic IR.
"""

import libadalang as lal
from lalcheck.ai.utils import profile
from lalcheck.tools.logger import log_stdout, log

import analysis
import typers
import utils
import time
from codegen import ConvertUniversalTypes, gen_ir


unknown_typer = typers.unknown_typer


class ExtractionContext(object):
    """
    The libadalang-based frontend interface. Provides method for extracting
    IR programs from Ada source files (see extract_programs), as well as
    a default typer for those programs (see default_typer).

    Note: programs extracted using different ExtractionContext are not
    compatible. Also, this extraction context must be kept alive as long
    as the programs parsed with it are intended to be used.
    """
    def __init__(self, lal_ctx=None):
        self.lal_ctx = lal.AnalysisContext() if lal_ctx is None else lal_ctx

        # Get a dummy node, needed to call static properties of libadalang.
        dummy = self.lal_ctx.get_from_buffer(
            "<dummy>", 'package Dummy is end;'
        ).root

        # Find the Character TypeDecl.
        char_type = dummy.p_standard_unit.root.find(
            lambda x: x.is_a(lal.TypeDecl) and x.f_name.text == "Character"
        )

        float_type = dummy.p_standard_unit.root.find(
            lambda x: x.is_a(lal.TypeDecl) and x.f_name.text == "Float"
        )

        self.evaluator = utils.ConstExprEvaluator(
            dummy.p_bool_type,
            dummy.p_int_type,
            float_type,
            char_type,
            dummy.p_universal_int_type,
            dummy.p_universal_real_type
        )

        self.type_models = {}
        self.fun_models = {}
        self.subpdata = {}
        self._internal_typer = self.default_typer()

    @staticmethod
    def for_project(project_file, scenario_vars={}):
        return ExtractionContext(lal.AnalysisContext(
            unit_provider=lal.UnitProvider.for_project(
                project_file,
                scenario_vars
            )
        ))

    @staticmethod
    def empty():
        return ExtractionContext()

    def extract_programs_from_file(self, ada_file):
        """
        :param str ada_file: A path to the Ada source file from which to
            extract programs.

        :return: a Basic IR Program for each subprogram body that exists in the
            given source code.

        :rtype: iterable[irt.Program]
        """
        return self._extract_from_unit(self.lal_ctx.get_from_file(ada_file))

    def extract_programs_from_provider(self, name, kind):
        return self._extract_from_unit(self.lal_ctx.get_from_provider(
            name, kind
        ))

    def extract_programs_from_unit(self, unit):
        """
        :param lal.AnalysisUnit unit: The already parsed compilation unit.
        """
        return self._extract_from_unit(unit)

    @profile()
    def use_model(self, name):
        model_unit = self.lal_ctx.get_from_provider(name, "specification")
        with log_stdout('error'):
            for diag in model_unit.diagnostics:
                print('   {}'.format(diag))
                return

        model_ofs = model_unit.root.findall(
            lambda x: (x.is_a(lal.AspectAssoc)
                       and x.f_id.text.lower() == "model_of")
        )

        type_models = {
            aspect.parent.parent.parent: aspect.f_expr.p_referenced_decl
            for aspect in model_ofs
            if aspect.parent.parent.parent.is_a(lal.TypeDecl, lal.SubtypeDecl)
        }

        fun_models = {
            aspect.parent.parent.parent: aspect.f_expr.p_referenced_decl
            for aspect in model_ofs
            if aspect.parent.parent.parent.is_a(lal.SubpDecl)
        }

        if len(type_models) + len(fun_models) != len(model_ofs):
            with log_stdout('info'):
                print("warning: detected usage of Model_Of in an unknown "
                      "context.")

        for tdecl, ref in type_models.iteritems():
            if ref is None:
                with log_stdout('info'):
                    print("warning: Model_Of used on '{}' refers to an unknown"
                          " type.".format(tdecl.f_name.text))
            else:
                self.type_models[ref] = tdecl

        for fdecl, ref in fun_models.iteritems():
            if ref is None:
                with log_stdout('info'):
                    print(
                        "warning: Model_Of used on '{}' refers to an unknown "
                        "procedure/function.".format(
                            fdecl.f_subp_spec.f_subp_name.text
                        )
                    )
            else:
                self.fun_models[ref] = fdecl

    @profile()
    def _extract_from_unit(self, unit):
        if unit.root is None:
            with log_stdout('error'):
                print('Could not parse {}:'.format(unit.filename))
                for diag in unit.diagnostics:
                    print('   {}'.format(diag))
                    return

        unit.populate_lexical_env()

        subpdata = analysis.traverse_unit(unit.root)
        self.subpdata.update(subpdata)

        progs = []
        for subp, subpuserdata in subpdata.iteritems():
            if subp.is_a(lal.BaseSubpBody):
                start_t = time.clock()
                progs.append(
                    gen_ir(self, subp, self._internal_typer, subpuserdata)
                )
                end_t = time.clock()
                log(
                    'timings',
                    " - Transformation of subprocedure {} took {}s".format(
                        subp.f_subp_spec.f_subp_name.text,
                        end_t - start_t
                    )
                )

        converter = ConvertUniversalTypes(self.evaluator, self._internal_typer)

        for prog in progs:
            prog.visit(converter)

        return progs

    def standard_typer(self):
        """
        :return: A Typer for Ada standard types of programs parsed using
            this extraction context.

        :rtype: types.Typer[lal.AdaNode]
        """
        return typers.standard_typer(self)

    def model_typer(self, inner_typer):
        return typers.model_typer(self, inner_typer)

    def default_typer(self, fallback_typer=None):
        """
        :return: The default Typer for Ada programs parsed using this
            extraction context.

        :rtype: types.Typer[lal.AdaNode]
        """
        return typers.default_typer(self, fallback_typer)
