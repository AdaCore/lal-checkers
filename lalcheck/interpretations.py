"""
Defines the TypeInterpreter interface, as well as a few common
TypeInterpreters.
"""

from lalcheck.domain_ops import (
    boolean_ops,
    interval_ops,
    finite_lattice_ops
)
from lalcheck.constants import ops, lits
from lalcheck import types
from lalcheck import domains


class TypeInterpreter(object):
    """
    A TypeInterpreter is used to determine how a type is interpreted in the
    backend. It must provide the domain used, a set of definitions that are
    available on this new domain, as well as a builder function to build
    elements of this domain from literal values.
    """
    def from_type(self, tpe):
        """
        Given a type, returns the domain that this TypeInterpreter would use to
        represent the type, a dictionary of definitions, as well as a builder
        for this domain. Returns None if this interpreter does not provide any
        interpretation for the given type.
        """
        raise NotImplementedError

    def __or__(self, other):
        """
        Creates a new interpreter by combining two interpreters. This new
        TypeInterpreter will try to interpret the given type using the first
        TypeInterpreter. It it failed (returned None), it will interpret it
        using the second TypeInterpreter.
        """
        @type_interpreter
        def f(hint):
            x = self.from_type(hint)
            return x if x is not None else other.from_type(hint)

        return f


def type_interpreter(fun):
    """
    A useful decorator to use on a function to turn it into an TypeInterpreter.
    The decorated function must receive a type as parameter and return
    a type interpretation.
    """
    class AnonymousInterpreter(TypeInterpreter):
        def from_type(self, tpe):
            return fun(tpe)

    return AnonymousInterpreter()


def dict_to_provider(def_dict):
    """
    Converts a dictionnary of definitions indexed by their names and domain
    signatures to a provider function.
    """
    def provider(name, signature):
        if (name, signature) in def_dict:
            return def_dict[name, signature]

    return provider


@type_interpreter
def default_boolean_interpreter(tpe):
    if tpe.is_a(types.Boolean):
        bool_dom = boolean_ops.Boolean
        un_fun_dom = (bool_dom, bool_dom)
        bin_fun_dom = (bool_dom, bool_dom, bool_dom)

        defs = {
            (ops.Not, un_fun_dom): boolean_ops.not_,
            (ops.And, bin_fun_dom): boolean_ops.and_,
            (ops.Or, bin_fun_dom): boolean_ops.or_,

            (ops.Eq, bin_fun_dom): finite_lattice_ops.eq(bool_dom),
            (ops.Neq, bin_fun_dom): finite_lattice_ops.neq(bool_dom)
        }

        inv_defs = {
            (ops.Not, un_fun_dom): boolean_ops.inv_not,
            (ops.And, bin_fun_dom): boolean_ops.inv_and,
            (ops.Or, bin_fun_dom): boolean_ops.inv_or,

            (ops.Eq, bin_fun_dom): finite_lattice_ops.inv_eq(bool_dom),
            (ops.Neq, bin_fun_dom): finite_lattice_ops.inv_neq(bool_dom)
        }

        builder = boolean_ops.lit

        return (
            bool_dom,
            dict_to_provider(defs),
            dict_to_provider(inv_defs),
            builder
        )


@type_interpreter
def default_int_range_interpreter(tpe):
    if tpe.is_a(types.IntRange):
        int_dom = domains.Intervals(tpe.frm, tpe.to)
        bool_dom = boolean_ops.Boolean
        unary_fun_dom = (int_dom, int_dom)
        binary_fun_dom = (int_dom, int_dom, int_dom)
        binary_rel_dom = (int_dom, int_dom, bool_dom)

        defs = {
            (ops.Plus, binary_fun_dom):
                interval_ops.add_no_wraparound(int_dom),

            (ops.Minus, binary_fun_dom):
                interval_ops.sub_no_wraparound(int_dom),

            (ops.Lt, binary_rel_dom): interval_ops.lt(int_dom),
            (ops.Le, binary_rel_dom): interval_ops.le(int_dom),
            (ops.Eq, binary_rel_dom): interval_ops.eq(int_dom),
            (ops.Neq, binary_rel_dom): interval_ops.neq(int_dom),
            (ops.Ge, binary_rel_dom): interval_ops.ge(int_dom),
            (ops.Gt, binary_rel_dom): interval_ops.gt(int_dom),
            (ops.Neg, unary_fun_dom): interval_ops.inverse(int_dom),
        }

        inv_defs = {
            (ops.Plus, binary_fun_dom):
                interval_ops.inv_add_no_wraparound(int_dom),

            (ops.Minus, binary_fun_dom):
                interval_ops.inv_sub_no_wraparound(int_dom),

            (ops.Lt, binary_rel_dom): interval_ops.inv_lt(int_dom),
            (ops.Le, binary_rel_dom): interval_ops.inv_le(int_dom),
            (ops.Eq, binary_rel_dom): interval_ops.inv_eq(int_dom),
            (ops.Neq, binary_rel_dom): interval_ops.inv_neq(int_dom),
            (ops.Ge, binary_rel_dom): interval_ops.inv_ge(int_dom),
            (ops.Gt, binary_rel_dom): interval_ops.inv_gt(int_dom),
            (ops.Neg, unary_fun_dom): interval_ops.inv_inverse(int_dom)
        }

        builder = interval_ops.lit(int_dom)

        return (
            int_dom,
            dict_to_provider(defs),
            dict_to_provider(inv_defs),
            builder
        )


@type_interpreter
def default_enum_interpreter(tpe):
    if tpe.is_a(types.Enum):
        enum_dom = domains.FiniteLattice.of_subsets(set(tpe.lits))
        bool_dom = boolean_ops.Boolean
        bin_rel_dom = (enum_dom, enum_dom, bool_dom)

        defs = {
            (ops.Eq, bin_rel_dom): finite_lattice_ops.eq(enum_dom),
            (ops.Neq, bin_rel_dom): finite_lattice_ops.neq(enum_dom)
        }

        inv_defs = {
            (ops.Eq, bin_rel_dom): finite_lattice_ops.inv_eq(enum_dom),
            (ops.Neq, bin_rel_dom): finite_lattice_ops.inv_neq(enum_dom)
        }

        builder = finite_lattice_ops.lit(enum_dom)

        return (
            enum_dom,
            dict_to_provider(defs),
            dict_to_provider(inv_defs),
            builder
        )


@type_interpreter
def simple_access_interpreter(tpe):
    if tpe.is_a(types.Pointer):
        ptr_dom = domains.FiniteLattice.of_subsets({lits.Null, lits.NotNull})
        bool_dom = boolean_ops.Boolean
        bin_rel_dom = (ptr_dom, ptr_dom, bool_dom)

        defs = {
            (ops.Eq, bin_rel_dom): finite_lattice_ops.eq(ptr_dom),
            (ops.Neq, bin_rel_dom): finite_lattice_ops.neq(ptr_dom)
        }

        inv_defs = {
            (ops.Eq, bin_rel_dom): finite_lattice_ops.inv_eq(ptr_dom),
            (ops.Neq, bin_rel_dom): finite_lattice_ops.inv_neq(ptr_dom)
        }

        builder = finite_lattice_ops.lit(ptr_dom)
        null = builder(lits.Null)
        notnull = builder(lits.NotNull)

        def def_provider(name, sig):
            if (name, sig) in defs:
                return defs[name, sig]

            elif (name == ops.Deref and
                    len(sig) == 2 and
                    sig[0] == ptr_dom):
                elem_dom = sig[1]

                def deref(ptr):
                    return elem_dom.bottom if ptr == null else elem_dom.top

                return deref

            elif (name == ops.Address and
                    len(sig) == 2 and
                    sig[1] == ptr_dom):
                elem_dom = sig[0]

                def address(elem):
                    return notnull

                return address

        def inv_def_provider(name, sig):
            if (name, sig) in defs:
                return inv_defs[name, sig]

            elif (name == ops.Deref and
                    len(sig) == 2 and
                    sig[0] == ptr_dom):
                elem_dom = sig[1]

                def inv_deref(elem, e_constr):
                    if ptr_dom.is_empty(e_constr) or elem_dom.is_empty(elem):
                        return None

                    if ptr_dom.le(notnull, e_constr):
                        return notnull

                    return None

                return inv_deref

            elif (name == ops.Address and
                    len(sig) == 2 and
                    sig[1] == ptr_dom):
                elem_dom = sig[0]

                def inv_address(ptr, e_constr):
                    if ptr_dom.is_empty(ptr) or elem_dom.is_empty(e_constr):
                        return None

                    return e_constr

                return inv_address

        return (
            ptr_dom,
            def_provider,
            inv_def_provider,
            builder
        )


default_type_interpreter = (
    default_boolean_interpreter |
    default_int_range_interpreter |
    default_enum_interpreter |
    simple_access_interpreter
)
