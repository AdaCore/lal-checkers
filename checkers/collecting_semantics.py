"""
An analyzer that collects semantics at each program point.
"""

from lalcheck.irs.basic.tools import (
    CFGBuilder,
    ExprEvaluator,
    ExprSolver
)

from lalcheck.irs.basic.tree import Variable
from lalcheck.irs.basic.tools import PrettyPrinter
from lalcheck.irs.basic.purpose import SyntheticVariable
from lalcheck.irs.basic import visitors
from lalcheck.utils import KeyCounter
from lalcheck.digraph import Digraph
from lalcheck import domains
from lalcheck import dot_printer

from xml.sax.saxutils import escape
from collections import defaultdict


class VarTracker(visitors.CFGNodeVisitor):
    def __init__(self, var_set, vars_domain, vars_idx, evaluator, c_solver):
        self.vars = var_set
        self.vars_domain = vars_domain
        self.vars_idx = vars_idx
        self.evaluator = evaluator
        self.constr_solver = c_solver

    def state_to_env(self, state):
        return {v: state[self.vars_idx[v]] for v in self.vars}

    def env_to_state(self, env):
        if len(env) == 0:
            return self.vars_domain.bottom
        else:
            return tuple(env[v] for v in self.vars)

    def visit_assign(self, assign, state):
        env = self.state_to_env(state)
        env[assign.id.var] = self.evaluator.eval(assign.expr, env)
        return self.env_to_state(env)

    def visit_assume(self, assume, state):
        return self.env_to_state(self.constr_solver.solve(
            assume.expr,
            self.state_to_env(state)
        ))

    def visit_read(self, read, state):
        return tuple(self.vars_domain.top[i] if i == self.vars_idx[read.id.var]
                     else x for i, x in enumerate(state))

    def visit_use(self, use, state):
        return state


class SimpleTraceLattice(domains.FiniteSubsetLattice):
    def __init__(self, *args):
        super(SimpleTraceLattice, self).__init__(*args)

    def update(self, a, b, widen=False):
        return super(SimpleTraceLattice, self).update(a, b, False)


class MergePredicateBuilder(object):
    def __init__(self, predicate):
        self.predicate = predicate

    def __or__(self, other):
        def f(trace_domain, vals_domain):
            s_pred = self.build(trace_domain, vals_domain)
            o_pred = other.build(trace_domain, vals_domain)

            return lambda a, b: s_pred(a, b) or o_pred(a, b)

        return MergePredicateBuilder(f)

    def __and__(self, other):
        def f(trace_domain, vals_domain):
            s_pred = self.build(trace_domain, vals_domain)
            o_pred = other.build(trace_domain, vals_domain)

            return lambda a, b: s_pred(a, b) and o_pred(a, b)

        return MergePredicateBuilder(f)

    def build(self, trace_domain, vals_domain):
        return self.predicate(trace_domain, vals_domain)


def _mp_always(*_):
    return lambda *_: True


def _mp_never(*_):
    return lambda *_: False


def _mp_le_traces(trace_domain, _):
    return lambda a, b: trace_domain.le(a[0], b[0])


def _mp_eq_vals(_, vals_domain):
    return lambda a, b: vals_domain.eq(a[1], b[1])


MergePredicateBuilder.Always = MergePredicateBuilder(_mp_always)
MergePredicateBuilder.Never = MergePredicateBuilder(_mp_never)
MergePredicateBuilder.Le_Traces = MergePredicateBuilder(_mp_le_traces)
MergePredicateBuilder.Eq_Vals = MergePredicateBuilder(_mp_eq_vals)


def html_render_node(node):
    return escape(PrettyPrinter.pretty_print(node))


def save_cfg_to(file_name, cfg):
    def render_node(node):
        return (html_render_node(node),) if node is not None else ()

    def render_widening_point(is_widening_point):
        return (escape('<widening_point>'),) if is_widening_point else ()

    with open(file_name, 'w') as f:
        f.write(dot_printer.gen_dot(cfg, [
            dot_printer.DataPrinter('node', render_node),
            dot_printer.DataPrinter('is_widening_point', render_widening_point)
        ]))


def repr_trace(trace):
    return tuple(p.name for p in trace)


def build_resulting_graph(file_name, cfg, results, trace_domain):
    paths = defaultdict(list)

    var_set = {
        v
        for node, state in results.iteritems()
        for trace, values in state.iteritems()
        for v, _ in values.iteritems()
    }

    for node, state in results.iteritems():
        for trace, values in state.iteritems():
            paths[frozenset(trace)].append(
                Digraph.Node(
                    node.name,
                    ___orig=node,
                    **{
                        v.name: value
                        for v, value in values.iteritems()
                        if not SyntheticVariable.is_purpose_of(v)
                    }
                )
            )

    edges = []
    for trace, nodes in paths.iteritems():
        for node in nodes:
            predecessors = [
                n
                for t, ns in paths.iteritems()
                for n in ns
                if trace_domain.le(t, trace)
                if n.data.___orig in cfg.ancestors(node.data.___orig)
            ]

            for pred in predecessors:
                edges.append(Digraph.Edge(pred, node))

    res_graph = Digraph(
        [n for _, nodes in paths.iteritems() for n in nodes],
        edges
    )

    def print_orig(orig):
        if orig.data.node is not None:
            return (
                '<i>{}</i>'.format(html_render_node(orig.data.node)),
            )
        return ()

    def print_result_builder(name):
        return lambda value: (name, str(value))

    with open(file_name, 'w') as f:
        f.write(dot_printer.gen_dot(res_graph, [
            dot_printer.DataPrinter('___orig', print_orig)
        ] + [
            dot_printer.DataPrinter(v.name, print_result_builder(
                "{} &isin;".format(v.name)
            ))
            for v in var_set
        ]))


class AnalysisResults(object):
    """
    Contains the results of the collecting semantics analysis.
    """
    def __init__(self, cfg, semantics, trace_domain, vars_domain,
                 evaluator):
        self.cfg = cfg
        self.semantics = semantics
        self.trace_domain = trace_domain
        self.vars_domain = vars_domain
        self.evaluator = evaluator

    def save_cfg_to_file(self, file_name):
        """
        Prints the control-flow graph as a DOT file to the given file name.
        """
        save_cfg_to(file_name, self.cfg)

    def save_results_to_file(self, file_name):
        """
        Prints the resulting graph as a DOT file to the given file name.
        Displays the state of each variable at each program point.
        """
        build_resulting_graph(
            file_name,
            self.cfg,
            self.semantics,
            self.trace_domain
        )

    def eval_at(self, node, expr):
        """
        Given a program point, evaluates for each program trace available at
        this program point the given expression.
        """
        return {
            trace: self.evaluator.eval(expr, env)
            for trace, env in self.semantics[node].iteritems()
        }


def collect_semantics(prog, model, merge_pred_builder):

    cfg = prog.visit(CFGBuilder())

    var_set = set(visitors.findall(prog, lambda n: isinstance(n, Variable)))
    vars_idx = {v: i for i, v in enumerate(var_set)}
    vars_domain = domains.Product(*(model[v].domain for v in var_set))
    trace_domain = SimpleTraceLattice(cfg.nodes)

    evaluator = ExprEvaluator(model)
    solver = ExprSolver(model)

    widening_counter = KeyCounter()
    widening_delay = 10

    def do_widen(counter):
        # will widen when counter == widen_delay, then narrow
        return counter == widening_delay

    do_stmt = VarTracker(var_set, vars_domain, vars_idx, evaluator, solver)

    lat = domains.Set(
        domains.Product(
            trace_domain,
            vars_domain
        ),
        merge_pred_builder.build(
            trace_domain,
            vars_domain
        ),
        None  # We don't need a top element here.
    )

    def it(states):
        new_states = states.copy()

        for node in cfg.nodes:
            inputs = [new_states[anc] for anc in cfg.ancestors(node)]
            res = (lat.build([(trace_domain.bottom, vars_domain.top)])
                   if len(inputs) == 0 else reduce(lat.join, inputs))

            res = lat.build([
                (
                    trace_domain.join(trace, trace_domain.build([node])),
                    updated_val
                )
                for trace, values in res
                for updated_val in [
                    node.data.node.visit(do_stmt, values)
                    if node.data.node is not None else values
                ]
                if not vars_domain.is_empty(updated_val)
            ])

            if node.data.is_widening_point:
                if do_widen(widening_counter.get_incr(node)):
                    res = lat.update(new_states[node], res, True)

            new_states[node] = res

        return new_states

    result = {n: lat.bottom for n in cfg.nodes}
    last = None
    while result != last:
        last, result = result, it(result)

    formatted_results = {
        node: {
            trace: {
                v: values[vars_idx[v]] for v in var_set
            }
            for trace, values in state
        }
        for node, state in result.iteritems()
    }

    return AnalysisResults(
        cfg,
        formatted_results,
        trace_domain,
        vars_domain,
        evaluator
    )
