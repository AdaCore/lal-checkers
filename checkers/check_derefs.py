import lalcheck.irs.basic.frontends.lal as lal2basic
import lalcheck.irs.basic.tree as irt
from lalcheck.irs.basic.tools import PrettyPrinter, Models
from lalcheck.irs.basic.purpose import DerefCheck
from lalcheck.digraph import Digraph
from lalcheck.interpretations import default_type_interpreter
from lalcheck.constants import lits
from lalcheck import dot_printer

from collecting_semantics import collect_semantics, MergePredicateBuilder

from xml.sax.saxutils import escape
from collections import defaultdict
import argparse


def html_render_node(node):
    return escape(PrettyPrinter.pretty_print(node))


def build_resulting_graph(file_name, cfg, null_derefs):
    def trace_id(trace):
        return str(trace)

    paths = defaultdict(list)

    for trace, derefed, precise in null_derefs:
        for node in trace:
            paths[node].append((trace, derefed, precise))

    new_node_map = {
        node: Digraph.Node(
            node.name,
            ___orig=node,
            **{
                trace_id(trace): (derefed, precise)
                for trace, derefed, precise in paths[node]
            }
        )
        for node in cfg.nodes
    }

    res_graph = Digraph(
        [new_node_map[n]
         for n in cfg.nodes],

        [Digraph.Edge(new_node_map[e.frm], new_node_map[e.to])
         for e in cfg.edges]
    )

    def print_orig(orig):
        if orig.data.node is not None:
            return (
                '<i>{}</i>'.format(html_render_node(orig.data.node)),
            )
        return ()

    def print_path_to_null_deref(value):
        derefed, precise = value
        qualifier = "" if precise else "potential "
        res_str = "path to {}null dereference of {}".format(
            qualifier, html_render_node(derefed)
        )
        return (
            '<font color="{}">{}</font>'.format('red', res_str),
        )

    with open(file_name, 'w') as f:
        f.write(dot_printer.gen_dot(res_graph, [
            dot_printer.DataPrinter('___orig', print_orig)
        ] + [
            dot_printer.DataPrinter(
                trace_id(trace),
                print_path_to_null_deref
            )
            for trace, _, _ in null_derefs
        ]))


class AnalysisResult(object):
    """
    Contains the results of the null dereference checker.
    """
    def __init__(self, sem_analysis, null_derefs):
        self.sem_analysis = sem_analysis
        self.null_derefs = null_derefs

    def save_results_to_file(self, file_name):
        """
        Prints the resulting graph as a DOT file to the given file name.
        At each program point, displays where the node is part of a path
        that leads to a (potential) null dereference.
        """
        build_resulting_graph(
            file_name,
            self.sem_analysis.cfg,
            self.null_derefs
        )


def check_derefs(prog, model, merge_pred_builder):

    analysis = collect_semantics(
        prog,
        model,
        merge_pred_builder
    )

    # Retrieve nodes in the CFG that correspond to program statements.
    nodes_with_ast = (
        (node, node.data.node)
        for node in analysis.cfg.nodes
        if 'node' in node.data
    )

    # Collect those that are assume statements and that have a 'purpose' tag
    # which indicates that this assume statement was added to check
    # dereferences.
    deref_checks = (
        (node, ast_node.data.purpose.expr)
        for node, ast_node in nodes_with_ast
        if isinstance(ast_node, irt.AssumeStmt)
        if DerefCheck.is_purpose_of(ast_node)
    )

    # Use the semantic analysis to evaluate at those program points the
    # corresponding expression being dereferenced.
    derefed_values = (
        (frozenset(trace) | {node}, derefed, value)
        for node, derefed in deref_checks
        for anc in analysis.cfg.ancestors(node)
        for trace, value in analysis.eval_at(anc, derefed).iteritems()
    )

    # Finally, keep those that might be null.
    # Store the program trace, the dereferenced expression, and whether
    # the expression "might be null" or "is always null".
    null_derefs = [
        (trace, derefed, len(value) == 1)
        for trace, derefed, value in derefed_values
        if lits.NULL in value
    ]

    return AnalysisResult(analysis, null_derefs)


def lal_subprogram_info(subp):
    return (
        subp.f_subp_spec.f_subp_name.text,
        subp.f_subp_spec.f_subp_name.sloc_range.start,
    )


def emit_codepeer_message(file, line, column,
                          proc_name, proc_file, proc_line, proc_column,
                          expr):
    print("{}:{}:{} warning: {}:{}:{}:{}: {} {}".format(
        file, line, column,
        proc_name, proc_file, proc_line, proc_column,
        "Null dereference of '{}'!".format(expr),
        "[check_derefs]"
    ))


def run(args):
    ctx = lal2basic.ExtractionContext()

    progs = ctx.extract_programs(args.file)

    model_builder = Models(
        ctx.default_typer(),
        default_type_interpreter
    )

    model = model_builder.of(*progs)

    if args.path_sensitive:
        merge_predicate = (
            MergePredicateBuilder.Le_Traces | MergePredicateBuilder.Eq_Vals
        )
    else:
        merge_predicate = MergePredicateBuilder.Always

    analyses = {
        prog: check_derefs(prog, model, merge_predicate)
        for prog in progs
    }

    if args.output_format == 'codepeer':
        emit_message = emit_codepeer_message

    for prog, analysis in analyses.iteritems():
        prog_info = lal_subprogram_info(prog.data.orig_node)
        for (trace, derefed, precise) in analysis.null_derefs:
            derefed_pos = derefed.data.orig_node.sloc_range.start
            emit_message(args.file, derefed_pos.line, derefed_pos.column,
                         prog_info[0], args.file,
                         prog_info[1].line, prog_info[1].column,
                         derefed.data.orig_node.text)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Null deref checker")
    parser.add_argument('--output-format', required=True, default="codepeer")
    parser.add_argument('--path-sensitive', action='store_true')
    parser.add_argument('file')

    run(parser.parse_args())
