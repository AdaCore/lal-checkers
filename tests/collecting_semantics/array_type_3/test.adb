procedure Ex1 is
   type My_Range is range 0 .. 20;
   type My_Enum is (A, B, C);
   type My_Array is array (My_Range, My_Enum) of Integer;
   x : My_Array;
begin
   x(12, B) := 14;
   x(4, A) := 12;
end Ex1;
