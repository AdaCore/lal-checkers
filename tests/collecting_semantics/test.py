import lalcheck.irs.basic.frontends.lal as lal2basic
from lalcheck.irs.basic.tools import Models
from lalcheck.interpretations import default_type_interpreter
from checkers.collecting_semantics import (
    collect_semantics,
    MergePredicateBuilder
)


def do(test_name, merge_predicate, idx=None):
    ctx = lal2basic.new_context()

    progs = lal2basic.extract_programs(
        ctx,
        '{}.adb'.format(test_name)
    )

    model_builder = Models(
        lal2basic.default_typer(ctx),
        default_type_interpreter
    )

    model = model_builder.of(*progs)

    collect_semantics(
        progs[0],
        model,
        merge_predicate,
        'output/{}_cfg.dot'.format(test_name),
        'output/{}{}_res.dot'.format(test_name, '' if idx is None else idx)
    )


le_trace_eq_vals = (
    MergePredicateBuilder.Le_Traces |
    MergePredicateBuilder.Eq_Vals
)

do('test_simple_if', MergePredicateBuilder.Always, '_always')
do('test_simple_if', le_trace_eq_vals, '_le_t_eq_v')

do('test_simple_if_expr', MergePredicateBuilder.Always, '_always')
do('test_simple_if_expr', le_trace_eq_vals, '_le_t_eq_v')

do('test_simple_while', le_trace_eq_vals)

do('test_simple_enum', MergePredicateBuilder.Always, '_always')

do('test_ptr', MergePredicateBuilder.Always, '_always')
do('test_ptr', le_trace_eq_vals, '_le_t_eq_v')
