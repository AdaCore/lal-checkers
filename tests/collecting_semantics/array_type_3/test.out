{
  "always": {
    "assign0": [
      {
        "trace:": [
          "assign0",
          "read0",
          "start0"
        ],
        "values": {
          "x": "[(((0, 11), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((13, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'A', u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'C'])), (14, 14))]"
        }
      }
    ],
    "assign1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "read0",
          "start0"
        ],
        "values": {
          "x": "[(((0, 11), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((13, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'A', u'C'])), (14, 14))]"
        }
      }
    ],
    "assign2": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "read0",
          "start0"
        ],
        "values": {
          "x": "[(((0, 11), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((13, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'A', u'C', u'B'])), (14, 14))]"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "read0",
          "start0"
        ],
        "values": {
          "x": "[(((0, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647))]"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "x": "[(((0, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647))]"
        }
      }
    ]
  },
  "le_t_eq_v": {
    "assign0": [
      {
        "trace:": [
          "assign0",
          "read0",
          "start0"
        ],
        "values": {
          "x": "[(((0, 11), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((13, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'A', u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'C'])), (14, 14))]"
        }
      }
    ],
    "assign1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "read0",
          "start0"
        ],
        "values": {
          "x": "[(((0, 11), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((13, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'A', u'C'])), (14, 14))]"
        }
      }
    ],
    "assign2": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "read0",
          "start0"
        ],
        "values": {
          "x": "[(((0, 11), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((13, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647)), (((12, 12), frozenset([])), (-2147483648, 2147483647)), (((12, 12), frozenset([u'A', u'C', u'B'])), (14, 14))]"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "read0",
          "start0"
        ],
        "values": {
          "x": "[(((0, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647))]"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "x": "[(((0, 20), frozenset([u'A', u'C', u'B'])), (-2147483648, 2147483647))]"
        }
      }
    ]
  }
}